075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,490 @@
+It means that the auto-rebalancing can start in the same reconciliation, being sure that the newly added pods are available for the upcoming rebalancing operation.
+Of course, the rebalancing will take time going through several reconciliations.
+
+For the scaling down, if the brokers to be removed are hosting topics' partitions, the operation is reverted before calling the `KafkaReconciler.reconcile()` method which is called with the original replicas count as if the cluster was never scaled down.
+If the brokers to be removed are not hosting any topics' partitions, the `KafkaReconciler.reconcile()` method will be called with the new replicas count and the scaling down is triggered by the `scaleDown()` call which updates the involved `StrimziPodSet`(s) so that the corresponding controller will take care of removing the brokers.
+In this case, the auto-rebalancing can start in the same reconciliation to take topics' partitions off but going through several reconciliations.
+When the auto-rebalancing ends, the triggered pods deletion takes some time but it doesn't have any impact on the rebalancing already done.
+Of course, if the scale down check doesn't fail, it means that no topics' partitions are hosted on the brokers to be removed and no auto-rebalancing is needed.
+
+The idea is about having a new `KafkaAutoRebalancingReconciler` class with its own `reconcile()` method which is called by the `KafkaAssemblyOperator` as one of the last steps in the reconciliation loop, so that:
+
+* we can get all the information about the brokers removed or added, which are available through the `KafkaClusterCreator` class (read later for more details).
+* we are sure that, in case of scaling up, the Cruise Control pod was restarted and it's now up and running again so it can accept new rebalancing requests.
+Of course the `KafkaAutoRebalancingReconciler.reconcile()` method also checks the status of an ongoing rebalancing across several reconciliations.
+
+It's also important taking into account that a user, by using `KafkaNodePool`(s), could remove and add brokers at the same time, so actually triggering both scaling down and up.
+In terms of scaling, the reconcile loop has the scale down triggered before the scaling up but it could be skipped because the brokers to be removed are still hosting topics' partitions. In such case the scaling up runs before.
+In terms of rebalancing, after the scaling up happens, the rebalancing for brokers to be removed takes the precedence on the rebalancing for newly added brokers. They are executed sequentially and not in parallel.
+
+### Auto-rebalancing Finite State Machine (FSM)
+
+The auto-rebalancing mechanism runs through a Finite State Machine (FSM) made by the following states:
+
+* **Idle**: Initial state with a new auto-rebalancing going to run because scaling down/up operations were requested. This is also the ending state after an auto-rebalancing completed successfully or failed.
```

Comment:
Will the `autoRebalance` in the Kafka CR status list the `Idle` state? Or will it be not present in the status and just handled as default?

Reply:
So, will the `Idle` state be removed after one reconciliation? It seems quite strange to have it not set on the beginning and always set afterwards to be honest.

Reply:
We could have it since the beginning, which would mean set it at the beginning of the reconciliation? But doesn't it trigger another reconciliation? Anyway, I would expect being immediately update to the next state, something like `RebalanceOnScaleXXX` at the end of the reconciliation when a rebalancing starts.

Reply:
Yeah, I think it would be more natural to set it there from the beginning. Or maybe even better from my perspective would be to not add the Idle state there at all and simply remove the field. But if that is a problem, having it there by default even before scaling makes more sense to me.

Reply:
Agreed. I will make clear it's set since the beginning.